routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,99 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Delegated Content Routing HTTP API
+
+**Author(s)**:
+- Gus Eggert
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+"Delegated content routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated routing.
+
+# API Specification
+The Delegated Content Routing Routing HTTP API uses the `application/json` content type by default. Clients and servers *should* support `application/cbor`, which can be negotiated using the standard `Accept` and `Content-Type` headers.
+
+## Common Data Types:
+
+- CIDs are always encoded using a [multibase](https://github.com/multiformats/multibase)-encoded [CIDv1](https://github.com/multiformats/cid#cidv1).
+- Multiaddrs are encoded according to the [human-readable multiaddr specification](https://github.com/multiformats/multiaddr#specification)
+- Peer IDs are encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use Base64.
+- Timestamps are Unix millisecond epoch timestamps
+
+## API
+- `GET /v1/providers/{CID}`
+    - Response
+        
+        ```json
+        {
+            "Providers": [
+                {
+					"Peer": {
+						"ID": "12D3K...",
+						"Multiaddrs": ["/ip4/.../tcp/.../p2p/...", "/ip4/..."]
```

Comment:
Doesn't really matter to me either way, but both `/ip4/1.2.3.4/tcp/1234` and `/ip4/1.2.3.4/tcp/1234/p2p/12D3...` are valid multiaddrs so I'm missing the distinction.

> is a directly serialized peer.AddrInfo

Are you looking for:
`"Peer"  : ["/ip4/.../tcp/.../p2p/12D3", "/ip4/.../quic/.../p2p/12D3", "/ip6/.../tcp/.../p2p/12D3",  ... ]` with the idea that the HTTP compression will reduce the overhead here?

Reply:
Nope, i'm just saying that if I `json.Marshal` a `peer.AddrInfo` the name we use for the addresses on that struct is `Addrs` not `Multiaddrs`, so if we use a different label here it would mean that we can't use the native type directly to represent a peer in golang.